File: Data Structures & Algorithms/invert-a-binary-tree/submission-10.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }


Questions
1. will we ever get an empty or single node tree

Approach -> one of the dfs algos
         -> return if current node is nul
		 -> swap Left child wtih Right chiuld
		 -> recurse Left or Right
		 -> time: o(n) -> where number of nodes is n
		 -> space: constant time, o(n) for call stack
 */

func invertTree(root *TreeNode) *TreeNode { 
	 //base case -> null tree
	 if root == nil{
		return nil
	 }

	 //recurse to children
	 invertTree(root.Left)
     invertTree(root.Right)

     //swap current node children
	 tmpNode := root.Right
	 root.Right = root.Left
	 root.Left = tmpNode
     
	 //return inverted tree
	 return root
}

File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-10.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        goodCount = 0
        q = deque()
        if root:
           q.append( (root, float("-inf")) )
         
        while len(q) > 0:
           node, maxPathVal = q.popleft()
           if node.val >= maxPathVal:
              goodCount += 1
              maxPathVal = node.val
           
           if node.left:
              q.append( (node.left, maxPathVal) )
           if node.right:
              q.append( (node.right, maxPathVal) ) 

        return goodCount
          